File: JS/Data Structures/Disjoint Set/DisjointSet.js
/**
 * A fast implementation of Disjoint-set data structure
 */
class DisjointSet {
  constructor() {
    /** @private */ this.parent = {};
    /** @private */ this.height = {};
    /** @private */ this.size = {};
    /** @private */ this.sets = 0;
  }

  /**
   * Add new node at the disjoint set
   * @param {*} node
   */
  add(node) {
    if (this.parent[node] === undefined) {
      this.parent[node] = node;
      this.height[node] = 0;
      this.size[node] = 1;
      this.sets += 1;
    }
  }

  /**
   * Returns the index of tree's root of the current node
   * @param {*}
   * @returns {*}
   */
  findSet(node) {
    let root = node;

    while (root !== this.parent[root]) {
      root = this.parent[root];
    }

    let tmp;
    while (root !== this.parent[node]) {
      tmp = this.parent[node];
      this.parent[node] = root;
      this.height[tmp] = 0;
      node = tmp;
    }

    return root;
  }

  /**
   * Find the index of tree's root of the node
   * @param {*} node - Node
   * @returns {*} the root of p
   */
  findSetRecursive(node) {
    if (node === this.parent[node]) {
      return node;
    }

    this.parent[node] = this.findSetRecursive(this.parent[node]);
    return this.parent[node];
  }

  /**
   * @param {*} nodeP
   * @param {*} nodeQ
   * @returns {boolean} true if p and q are in the same set, false in otherwise
   */
  isSameSet(nodeP, nodeQ) {
    return this.findSet(nodeP) === this.findSet(nodeQ);
  }

  /**
   * Join the trees that contains the nodes
   * @param {*} nodeP
   * @param {*} nodeQ
   */
  unionSet(nodeP, nodeQ) {
    const rootP = this.findSet(nodeP);
    const rootQ = this.findSet(nodeQ);

    if (rootP !== rootQ) {
      if (this.height[rootP] < this.height[rootQ]) {
        this.parent[rootP] = rootQ;
        this.size[rootQ] += this.size[rootP];
      } else {
        this.parent[rootQ] = rootP;
        this.size[rootP] += this.size[rootQ];
        if (this.height[rootP] === this.height[rootQ]) {
          this.height[rootP] += 1;
        }
      }
      this.sets -= 1;
      return true;
    }

    return false;
  }

  /**
   * @param {*} node
   * @return {number}
   */
  sizeOfSet(node) {
    const rootNode = this.findSet(node);
    return this.size[rootNode];
  }

  /**
   * @return {number}
   */
  numSets() {
    return this.sets;
  }


  /**
   * @override
   * @returns {string}
   */
  toString() {
    const output = [];
    for (const node in this.parent) {
      output.push(`[${node} -> ${this.parent[node]}]`);
    }
    return output.join(', ');
  }
}

module.exports = DisjointSet;

File: C++/Data Structures/DisjointSet.cpp
#include <iostream>
#include <map>

using namespace std;

template <class T> class DisjointSet {
  private:
    map<T,T> parent;
    map<T,int> height;
    map<T,int> size;
    int sets;

  public:
    DisjointSet() {
      parent = map<T,T>();
      height = map<T,int>();
      size   = map<T,int>();
      sets = 0;
    }

    void add(T node) {
      if (parent.find(node) == parent.end()) {
        parent[node] = node;
        height[node] = 0;
        size[node]   = 1;
        sets        += 1;
      }
    }

    T findSet(T node) {
      T root = node;
      while (root != parent[root]) {
        root = parent[root];
      }

      T tmp;
      while (root != parent[node]) {
        tmp = parent[node];
        parent[node] = root;
        height[tmp] = 0;
        node = tmp;
      }

      return root;
    }

    bool isConnected(T nodeP, T nodeQ) {
      return findSet(nodeP) == findSet(nodeQ);
    }

    bool unionSet(T nodeP, T nodeQ) {
      T rootP = findSet(nodeP);
      T rootQ = findSet(nodeQ);

      if (rootP != rootQ) {
        if(height[rootP] < height[rootQ]) {
          parent[rootP] = rootQ;
          size[rootQ] += size[rootP];
        } else {
          parent[rootQ] = rootP;
          size[rootP] += size[rootQ];
          if (height[rootP] == height[rootQ]) {
            height[rootP]++;
          }
        }
        sets--;
        return true;
      }
      return false;
    }

    int sizeOfSet(T node) {
      T rootNode = findSet(node);
      return size[rootNode];
    }

    int numOfSets() {
      return sets;
    }
};

int main()
{

    return 0;
}
